ttnn/cpp/ttnn/graph_processor.cpp

```diff
@@ -0,0 +1,450 @@
+
+    }
+    void GraphProcessor::track_allocate(tt::tt_metal::Buffer* buffer, bool bottom_up) {
+        const std::lock_guard<std::mutex> lock(mutex);
+        auto buf_id = add_buffer(buffer);
+
+        auto alloc_id = reinterpret_cast<std::uintptr_t>(buffer);
+        auto counter = graph.size();
+
+        std::unordered_map<std::string, std::string> params = {
+                {"size", std::to_string(buffer->size())},
+                {"address", std::to_string(buffer->address())},
+                {"type", buffer->is_dram() ? "DRAM" : "L1"},
+                {"layout", tensorMemoryLayoutToString(buffer->buffer_layout())}
+        };
+        {
+            graph.push_back(Vertex{
+                .counter = counter,
+                .name = "buffer_allocate",
+                .params = params,
+                .connections = {buf_id}
+            });
+            graph[current_op_id.top()].connections.push_back(counter);
+        }
+    }
+
+    void GraphProcessor::track_deallocate(tt::tt_metal::Buffer* buffer) {
+        const std::lock_guard<std::mutex> lock(mutex);
+        auto counter = graph.size();
+        auto buffer_idx = add_buffer(buffer);
+        std::unordered_map<std::string, std::string> params = {
+                {"size", std::to_string(buffer->size())},
+                {"type", buffer->is_dram() ? "DRAM" : "L1"},
+                {"layout", tensorMemoryLayoutToString(buffer->buffer_layout())}
+        };
+        {
+            graph.push_back(Vertex{
+                .counter = counter,
+                .name = "buffer_deallocate",
+                .params = params,
+                .connections = {buffer_idx}
+            });
+            graph[current_op_id.top()].connections.push_back(counter);
+        }
+
+    }
+
+    void GraphProcessor::track_allocate_cb(const CoreRangeSet &core_range_set, uint64_t addr, uint64_t size) {
+        const std::lock_guard<std::mutex> lock(mutex);
+        if (addr == 0 && ! fake_run) {
+            return;
+        }
+        std::unordered_map<std::string, std::string> params = {
+            {"size", std::to_string(size)},
+            {"addr", std::to_string(addr)},
+            {"core_range_set", core_range_set.str()}
+        };
+        auto counter = graph.size();
+        {
+            graph.push_back({
+                .counter = counter,
+                .name = "circular_buffer_allocate",
+                .params = params,
+                .connections = {}
+            });
+            graph[current_op_id.top()].connections.push_back(counter);
+        }
+
+    }
+
+    void GraphProcessor::track_deallocate_cb() {
+        const std::lock_guard<std::mutex> lock(mutex);
+        auto counter = graph.size();
+        {
+            graph.push_back(Vertex{
+                .counter = counter,
+                .name = "circular_buffer_deallocate_all",
```

Comment:
should this be?
```suggestion
                .name = "circular_buffer_deallocate",
```

Reply:
we deallocate all at once. so I want it to differ from circular_buffer_allocate